### DISCUSSION TITLE
Modification in Action kernel

### DISCUSSION THREAD

**Discussion by ashishdhole** [idaholab/moose#27374] [Q&A Modules: Phase field]
Hello, 
I want to know how Action kernel works. In grain growth model, GrainGrowthAction takes care of time derivative, Allen Cahn and interface. Now interface (ACInterface) kernel is having negative sign in the discription, but I don't see that negative sign in ACInterface or in Action kernel. If I want to add an extra term to my PDE that is being subtracted from the Allen Cahn kernel (ACGrGrPoly in this case) should I include (-) sign in my new kernel in both the residual and jacobian part or I need to modify the Actions kernel. 
Thank you

**Comment by GiudGiud** [comment]
Hello

An Action is something different from a kernel. An action can add objects to the simulation, such as kernels. 
You should not need to modify exisiting kernels to add another term to the equation. You can just create a new kernel if we have not implemented one that fits what you need

Guillaume

**Reply by ashishdhole** [reply]
Hello Guillaume, 
I am trying to incorporate drag pressure into the grain growth model, This require an extra term in the PDE that resist the growth. for which I made a kernel. so my question in other words is, if my action describes A+B-C=0 (thats the grain growth model, A=time derivative, B=allen cahn and c= interface kernel), if I want to add/subtract D kernel to action, can I not just add the block that discribes A+B-C+D=0.

**Reply by GiudGiud** [reply]
You can add a kernel in the `[Kernels]` block even if the Action is in use.

**Reply by ashishdhole** [reply]
adding extra kernel to input file need v and variable on which it is acted on. But for a system of say 100 order parameters, which should be the variable. The module/action part usually tkes care of such variable assignment

**Reply by GiudGiud** [reply]
if you want it to be automatic, then definitely do it in the action instead. This will take a little C++, it should be reasonable though., about as difficult as creating a kernel

**Reply by ashishdhole** [reply]
great! 
Now I have made the kernel and I wanted to add it to the actions. But I have a doubt. In grain growth model the ACInterface is having a -ve sign in its weak form (As expressed on the website) but I don't see any such sign in the GrainGrowthAction nor do I see it in the ACInterface Kernel. Am I missing anytihng? I want to add a tern that is subtracted from AllenCan i.e. ACGrGrPoly kernel in GrainGrowthAction.

**Reply by GiudGiud** [reply]
The sign may have been flipped because the term is moved to the left hand side?
For example if you look at Reaction.C, it's a loss term (should have a minus) but it s moved to the LHS so there's a plus in the computeQpResidual

**Reply by ashishdhole** [reply]
But I see +ve in both ACGrGrPoly and in ACKernel if they are used in same action (GrainGrowthAction). So computeQpResidual for both these kernels shows +ve sign, which according to the equation should be opposite. is there any term/variable that is added/introduced to the ACInterface kernel which takes care of the -ve sign? Trying to figure out how actions works. I see TimeDerivative as +ve as well, which makes sense because it is in LHS from the beginning, ACGrGrPoly also should be +ve as it is -L(df/dn) in the strong form, but I am not getting why ACInterface should be +ve as well. can you please explain?

**Reply by GiudGiud** [reply]
@laagesen 

I can help with MOOSE. I'd rather call in an expert for phase field

**Comment by ashishdhole** [comment]
Sure, Thank you Guillaume.

**Comment by laagesen** [comment]
Hi, ACInterface should have a positive sign, this is correct. The one step you didn't mention that introduces a change in sign is using the product rule/integration by parts to reduce the order of derivative on \eta in the ACInterface kernel: the product rule is often written in a simple way udv = uv - vdu, so you can see the sign change there. I think this is the step you are missing. This page gives a step by step description of how the residual equation is arrived at:

https://mooseframework.inl.gov/modules/phase_field/Phase_Field_Equations.html

If you're not familiar with the product rule/integration by parts to reduce the order of derivative as is often done in FEM discretization, I'd suggest reviewing the MOOSE framework slides and on-line training, I think it is covered pretty well there.

**Reply by ashishdhole** [reply]
Is there any example that shows the usage of additional functional density (f_add) as explained in grain growth model? I have one term in the evolution PDE that I need to subtract from f_loc.

**Comment by laagesen** [comment]
We don't currently have a way to do this in an automated way, unfortunately, but the `PolycrystalKernelAction` could be modified to add this.

**Reply by ashishdhole** [reply]
I actually made another Action by adding the new kernel to it and it works (kind of) but the behaviour is very different than what I would expect. Say for example, I an making a kernel that is to be subtracted from f_loc, will my residue, jacobian and offdiagonal jacobian will have negative sign in it?  Can I in some way call the residual of another kernel in my new kernal as a condition. Say if residual of ACGrGrPoly is > 0, the new kernel residual be X and if it is < 0 the residual be Y? Or do I have to integrate my physics and the existing one in one kernel to get it done?

**Reply by laagesen** [reply]
Can you say a little more about what physics you are trying to incorporate in f_loc? I can't think of an example where you would want to subtract, rather than add, energy to the system.

**Reply by ashishdhole** [reply]
I want to subtract (3*eta_i*eta_j*P_z) from f_loc but under some condition.

**Reply by ashishdhole** [reply]
3 * eta_i * eta_j * P_z

**Reply by ashishdhole** [reply]
It is a constant friction pressue that depicts a pinning pressure without actually putting a particle in the simulation. I am following Shahandeh, S., M. Greenwood, and Matthias Militzer. "Friction pressure method for simulating solute drag and particle pinning in a multiphase-field model." Modelling and Simulation in Materials Science and Engineering 20, no. 6 (2012): 065008. for drag pressure model.

**Reply by laagesen** [reply]
I'd suggest adding a material property that is a prefactor that multiplies 3eta_ieta_j*P_z to the new kernel. Then you can create a material property using the DerivativeParsedMaterial system that is -1 when you meet the condition you want to subtract and 0 when you don't want to subtract. This will be easier and more flexible than trying to check what is going on in a different kernel from within your new kernel.

**Reply by ashishdhole** [reply]
Okay. Do we have any examples of similar type that I can go through to implement it?

**Reply by laagesen** [reply]
MatDiffusion or MatReaction both have examples of material property multiplying the other terms in the kernel.

**Reply by ashishdhole** [reply]
Great! Thank you for your help on this. Let me try it.

**Reply by laagesen** [reply]
MatReaction is probably better to start looking at, because MatDiffusion is templated to allow for either scalar or tensor diffusivities. Good luck!
